### PR TITLE
Add promise-like interface

### PR DESCRIPTION
**Pull request by zhujinxuan**
## Description
https://github.com/ant-design/ant-design/issues/10421

This PR provides the then(onClose) promise-like interface for the message

## Check List
Please makes sure that these checkboxes are checked before submitting your PR, thank you!

* [x] Make sure that you propose PR to right branch: bugfix for `master`, feature for latest active branch `feature-x.x`.
* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [x] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [x] Rebase before creating a PR to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you PR.

Extra checklist:

**elif** *isNewFeature* **:**

  * [x] Update API docs for the component.
  * [x] Update Demo for new feature
  * [x] Update TypeScript definition for the component.
  * [x] Add unit tests for the feature.

### CODE REVIEW

File: components/message/index.tsx [hunk 1]
```diff
@@ -54,25 +64,36 @@ function notice(
...
-      duration,
-      style: {},
-      content: (
-        <div className={`${prefixCls}-custom-content ${prefixCls}-${type}`}>
-          <Icon type={iconType} />
-          <span>{content}</span>
-        </div>
-      ),
-      onClose,
+  const closePromise = new Promise((resolve) => {
```
Comment (yesmeck) [inline components/message/index.tsx]: https://caniuse.com/#search=Promise

File: components/message/__tests__/index.test.js [hunk 2]
```diff
@@ -73,16 +73,15 @@ describe('message', () => {
...
     const now = Date.now();
-    message.info('whatever', () => {
+    return message.info('whatever', () => {
       // calculate the approximately duration value
       const aboutDuration = parseInt((Date.now() - now) / 1000, 10);
       expect(aboutDuration).toBe(defaultDuration);
-      done();
-    });
+    }).then(() => true);
   });
```
Comment (yesmeck) [inline components/message/__tests__/index.test.js]: Write a separator test case.

File: components/message/index.tsx (hunk 1 above)
Comment (zhujinxuan) [reply components/message/index.tsx]: I am not sure.  Do you mean some browsers ant-design supporting does not support `Promise`?

File: components/message/__tests__/index.test.js (hunk 2 above)
Comment (zhujinxuan) [reply components/message/__tests__/index.test.js]: Fixed.  Thank you.

File: components/message/index.tsx (hunk 1 above)
Comment (yesmeck) [reply components/message/index.tsx]: I think it's ok, never mind.

File: components/message/index.zh-CN.md [hunk 3]
```diff
@@ -35,6 +35,12 @@ title: Message
 - `message.config(options)`
 - `message.destroy()`
 
+组件同时提供 then 接口
```
Comment (afc163) [inline components/message/index.zh-CN.md]: then 改成 promise，后面加句号。

Comment (afc163) [reply components/message/index.zh-CN.md]: 加一个空行。

File: components/message/index.zh-CN.md [hunk 4]
```diff
@@ -35,6 +35,12 @@ title: Message
...
+- `message[level](content, [duration] ).then(afterClose)`
+- `message[level](content, [duration], onClose ).then(afterClose)`
```
Comment (afc163) [inline components/message/index.zh-CN.md]: onClose 后面多余空格。

File: components/message/index.zh-CN.md [hunk 5]
```diff
@@ -35,6 +35,12 @@ title: Message
...
+
+其中`message[level]` 是组件已经提供的静态方法。`then` 接口返回值是 Promise
```
Comment (afc163) [inline components/message/index.zh-CN.md]: `message 前端少了一个空格。

Comment (afc163) [reply components/message/index.zh-CN.md]: 句尾加标点。

**Comment by afc163** [comment]
再补充一下用例吧。

**Comment by zhujinxuan** [comment]
@afc163 请问建议补充怎样的用例呢？

**Comment by afc163** [comment]
测试 promise 用法有效并符合预期。

**Comment by zhujinxuan** [comment]
@afc163 请问你指的是这个 PR 中的 `components/message/demo/thenable.md` 么？
我刚刚修改了这个的文档，让这个文档更加明了一些

File: components/message/demo/loading.md [hunk 6]
```diff
@@ -19,7 +19,7 @@ import { message, Button } from 'antd';
 const success = () => {
   const hide = message.loading('Action in progress..', 0);
   // Dismiss manually and asynchronously
-  setTimeout(hide, 2500);
+  setTimeout(hide, 500);
```
[the remark this answers is not in the document]
Comment (zhujinxuan) [reply components/message/demo/loading.md]: I revert the change in recent commit 799bcfbac.  Thank you.

**Pull request merged**